### ISSUE TITLE
ENH: automatically use expand_dims for concatenating along axis not contained in arrays

### ISSUE THREAD

**Issue by h-vetinari** [numpy/numpy#11827]
### Reproducing code example:

I'm surprised that the following does not work out of the box (couldn't find an existing issue for it), as I think the intent is unambiguous:
```python
import numpy as np
a = np.zeros((3,))
b = np.ones((3,))
np.concatenate([a, b], axis=0) # works
# array([0., 0., 0., 1., 1., 1.])
c = np.concatenate([a, b], axis=1)
# AxisError: axis 1 is out of bounds for array of dimension 1
```

This is of course a very simplified example, in practice, `a/b` (or many more) get produced somewhere else, and I need to calculate on `c`. There's at least two work-arounds (probably many more), which are not particularly elegant:

```
np.concatenate([np.expand_dims(x, 1) for x in [a, b]], axis=1) # verbose
# array([[0., 1.],
#        [0., 1.],
#        [0., 1.]])
np.concatenate([[a], [b]], axis=0).T # hard to follow
# array([[0., 1.],
#        [0., 1.],
#        [0., 1.]])
```

Within the larger context of different dimensionalities, I think `concatenate` should (after checking equality of `shape` for the array sequence, of course) just default to `expand_dims` if `axis >= len(arr.shape)`, conceptually something like:

```
# call to: np.concatenate(arr, axis=ax)
[...]
s = arr[0].shape
for i in range(1, len(arr)):
    assert s == arr[i].shape
if ax >= len(s):
    arr = [np.expand_dim(a, ax) for a in arr]
# actually concatenate
[...]
```

### Numpy/Python version information:

```1.15.0, 3.6.6```

**Comment by eric-wieser** [comment]
I think you're looking for `np.stack`, which does pretty much what you're asking for.

**Comment by h-vetinari** [comment]
Thanks for the tip! However, I still think the current behaviour is not very consistent:

```
a = np.zeros((3,))
b = np.ones((3,))

np.concatenate([a, b], axis=0)
# array([0., 0., 0., 1., 1., 1.])

np.concatenate([a, b], axis=1) # the only error of the bunch!
# AxisError: axis 1 is out of bounds for array of dimension 1

np.stack([a, b], axis=1)
# array([[0., 1.],
#        [0., 1.],
#        [0., 1.]])

np.stack([a, b], axis=0)
# array([[0., 0., 0.],
#        [1., 1., 1.]])

np.hstack([a, b])
# array([0., 0., 0., 1., 1., 1.])

np.vstack([a, b])
# array([[0., 0., 0.],
#        [1., 1., 1.]])

np.dstack([a, b])
# array([[[0., 1.],
#         [0., 1.],
#         [0., 1.]]])
```

Plus, `np.stack` inserts another axis instead of using the existing one in case the desired `axis` *does* exist within the arrays, which I find much less intuitive than `concatenate`. From the docs (https://docs.scipy.org/doc/numpy/reference/generated/numpy.stack.html):

```
arrays = [np.random.randn(3, 4) for _ in range(10)]
np.stack(arrays, axis=0).shape
# (10, 3, 4)
np.stack(arrays, axis=1).shape
# (3, 10, 4)
np.stack(arrays, axis=2).shape
# (3, 4, 10)
```

I still think this could be a reasonable enhancement, not least for people coming from `pandas`, where `pd.concat` does handle the case I describe.

**Comment by seberg** [comment]
As a gut feeling, I have a little problem with this, since especially if the inputs have different shapes (e.g. the axis sometimes exists), you are suggesting something that is not compatible with normal broadcasting.
So, I am a bit doubtful about whether it is obvious enough what happens to be honest. Pandas is a bit different probably, since it is not really N-D (although maybe I am missing the specific use case in pandas).
Some of the stacking functions actually have this functionality, never liked it much, but honestly, I don't think I ever really have a mixed dimension case.

**Comment by eric-wieser** [comment]
> I still think the current behaviour is not very consistent:

Sadly, `hstack`, `vstack` and `dstack` are far from consistent.

If you want consistency, I'd stick with the more primitive `concatenate` (which combines along an existing axis) or `stack` (which concatenates along a new axis).

[In the face of ambiguity, refuse the temptation to guess](https://zen-of-python.info/in-the-face-of-ambiguity-refuse-the-temptation-to-guess.html#12) and all that - if you want to insert extra dimensions, then I'd prefer you be explicit about it, rather than creating an API that guesses where you wanted them

**Comment by h-vetinari** [comment]
@seberg 
> [...] especially if the inputs have different shapes [...]

I didn't suggest that, in fact I wrote (also in the pseudocode at the end):
> (after checking equality of `shape` for the array sequence, of course)

**Comment by h-vetinari** [comment]
@eric-wieser 
> In the face of ambiguity, refuse the temptation to guess and all that - if you want to insert extra dimensions, then I'd prefer you be explicit about it, rather than creating an API that guesses where you wanted them.

That's fair enough, but we might disagree about what's ambiguous. To me,
```
a = np.zeros((3,))
b = np.ones((3,))
np.concatenate([a, b], axis=0) # works
# array([0., 0., 0., 1., 1., 1.])
np.concatenate([a, b], axis=1) # hypthetical!
# array([[0., 1.],
#        [0., 1.],
#        [0., 1.]])
```
**is** unambiguous _and_ consistent. Once I'm asking to concatenate along `axis=0`, and once along `axis=1` - which does not exist in the arrays themselves, but makes perfect sense.

**Comment by eric-wieser** [comment]
What should `np.concatenate([a, b], axis=2)` do under your rules? How about `np.concatenate([a, b], axis=-2)`?

**Comment by h-vetinari** [comment]
@eric-wieser 
To me, `np.concatenate([a, b], axis=i)` should resolve to (after the shape checks)
```np.concatenate([np.expand_dims(x, axis=i) for x in [a, b]], axis=i)```

I see now that `expand_dims` has deprecated adding more than one dimension -- which is honestly all that's needed for `concatenate` -- so I see no harm in restricting my proposal to that.

**Comment by eric-wieser** [comment]
> I see now that expand_dims has deprecated adding more than one dimension

Did it ever support that?

**Comment by eric-wieser** [comment]
> should resolve to ... `np.concatenate([np.expand_dims(x, axis=i) for x in [a, b]], axis=i)`

This is exactly the definition of `np.stack`

**Comment by h-vetinari** [comment]
> Did it ever support that?

Not familiar with its history, never used it before a day ago. But it's easy to try:
```
a = np.zeros((3,))
b = np.ones((3,))
[np.expand_dims(x, axis=2) for x in [a, b]]
# DeprecationWarning: Both axis > a.ndim and axis < -a.ndim - 1 are deprecated and will raise an AxisError in the future.
  
# [array([[0.],
#         [0.],
#         [0.]]), array([[1.],
#         [1.],
#         [1.]])]
```

**Comment by eric-wieser** [comment]
Note that the output is wrong though - you asked to insert a dimension at axis 2 (ie make 3d), but it produced a 2D result. The output didn't change - all that did is that getting the wrong output is deprecated.

**Comment by eric-wieser** [comment]
> is the only one of the bunch to throw an error 

Well, to be fair you didn't pass an `axis` argument to anything else other than `stack` and `concatenate`, so really we can only compare those.

If you passed `axis=2`, you'd find both fail. For N-dimensional arrays (and ignoring negatives):
* `stack` takes an axis in `range(0, N+1)` - the location of the new axis in the destination array
* `concatenate` takes an axis in `range(0, N) - the location of the combined axis in the destination array

> using np.stack .. yields a counter-intuitive result for me - inserting an axis

I think it's worth consulting the documentation here for these two functions. The one-line summaries are:
```
concatenate : Join a sequence of arrays along an existing axis.
stack       : Join a sequence of arrays along a new axis.
```

Perhaps the names "stack" and "concatenate" don't convey this well, but we can't change them easily now, and off the top of my head, I can't come up with anything nearly as concise that helps.

**Comment by seberg** [comment]
@h-vetinari I see that you wanted to use `expand_dims`. If you do not want it to work when the ndim differs (i.e. some are expanded and some are not), then `np.stack` seems to do *exactly* what you want.

Now, the next point seems to be that you want to switch between the two behaviours based on the input dimension. Frankly, that just seems unnecessary? You typically should know the dimension of the input already, even in an interactive session I am not sure that switching between the two commands is much of a hassle. And automagically inserting the axis in `concatenate` could also lead to not noticing missing input validation. Sure, you probably know that you really want to add that axis if it does not exist, but in that case spelling out does not seem so bad and accidentally adding it can be inconsistent as well.

Frankly, I will close the issue for now. I think there are at least two core devs who feel this is too far on the side losing consistency for a bit of quick convenience. You could bring it up on the mailing list to rally a bit support.

**Comment by h-vetinari** [comment]
@eric-wieser 
> I think it's worth consulting the documentation here for these two functions. The one-line summaries are:
  ```
  concatenate : Join a sequence of arrays along an existing axis.
  stack       : Join a sequence of arrays along a new axis.
  ```

I think this could reasonably be changed to 
```
concatenate : Join a sequence of arrays along a given axis.
stack       : Join a sequence of arrays along an inserted axis.
```
This seems clearer (would also explain ** in https://github.com/numpy/numpy/issues/11827#issuecomment-416852036 more lucidly) IMO, and would allow for both to have `range(0, ndim+1)` for axis, with only the case `axis=ndim+1` coinciding between the two.

@seberg 
> @h-vetinari I see that you wanted to use expand_dims. If you do not want it to work when the ndim differs (i.e. some are expanded and some are not), then np.stack seems to do exactly what you want.

My point is that `concatenate` should be consistent across `range(0, ndim+1)` - it's the go-to function to concatenate things, and concatenating vectors into matrices (or similar in higher dimension) is a very often occurring operation *that should not necessitate the switch* from `concatenate` to `stack` for `axis=1`.

> Frankly, I will close the issue for now. I think there are at least two core devs who feel this is too far on the side losing consistency for a bit of quick convenience. You could bring it up on the mailing list to rally a bit support.

No skin off my back that you close it, but for the record, I strongly disagree with "losing consistency for a bit of quick convenience".

**Comment by billtubbs** [comment]
This discussion was interesting and mostly over my head but I came here looking for a way to horizontally stack a mixture of 1 and 2-d arrays, which I think is a common special case of what the original person was asking.

For this there is the aptly-named [column_stack](https://numpy.org/doc/stable/reference/generated/numpy.column_stack.html) function:
```
a = np.zeros((3,))
b = np.ones((3,))
c = np.arange(9).reshape((3, 3))
np.column_stack([a, b, c])

# array([[0., 1., 0., 1., 2.],
#        [0., 1., 3., 4., 5.],
#        [0., 1., 6., 7., 8.]])
```
